k8s/src/main/scala/io/buoyant/k8s/ServiceNamer.scala

```diff
@@ -24,26 +23,135 @@ class ServiceNamer(
+        logEvent.addition(svc.ports -- ports.keys)
+        svc
+      case v1.ServiceModified(service) =>
+        val svc = Svc(service)
+        logEvent.addition(svc.portMappings -- portMappings.keys)
+        logEvent.addition(svc.ports -- ports.keys)
+        logEvent.deletion(portMappings -- svc.portMappings.keys)
+        logEvent.deletion(ports -- svc.ports.keys)
+        logEvent.modification(portMappings, svc.portMappings)
+        logEvent.modification(ports, svc.ports)
+        svc
+      case v1.ServiceDeleted(deleted) =>
+        val Svc(deletedPorts, deletedMappings) = Svc(deleted)
+        logEvent.deletion(deletedPorts)
+        logEvent.deletion(deletedMappings)
+        this.copy(
+          ports = ports -- deletedPorts.keys,
+          portMappings = portMappings -- deletedMappings.keys
+        )
+      case v1.ServiceError(error) =>
+        log.warning("k8s ns %s service %s error %s", logEvent.nsName, logEvent.serviceName, error)
+        this
+    }
+  }
+
+  private[this] object Svc {
+    def apply(service: v1.Service): Svc = {
+      val ports = mutable.Map.empty[String, Address]
+      val portMap = mutable.Map.empty[Int, String]
+
+      for {
+        meta <- service.metadata.toSeq
+        name <- meta.name.toSeq
+        status <- service.status.toSeq
+        lb <- status.loadBalancer.toSeq
+        spec <- service.spec.toSeq
+        port <- spec.ports
+      } {
+        for {
+          ingress <- lb.ingress.toSeq.flatten
+          hostname <- ingress.hostname.orElse(ingress.ip)
+        } ports += port.name -> Address(new InetSocketAddress(hostname, port.port))
+
+        portMap += (port.targetPort match {
+          case Some(targetPort) => port.port -> targetPort
+          case None => port.port -> port.port.toString
+        })
+      }
+      Svc(ports.toMap, portMap.toMap)
+    }
+    def fromResponse(response: Option[v1.Service]): Svc =
+      response.map(Svc(_)).getOrElse(Svc(Map.empty, Map.empty))
+  }
 
-    case (id@Path.Utf8(nsName, portName, serviceName, labelValue), Some(label)) =>
-      val nameTree = serviceNs
-        .get(nsName.toLowerCase, Some(s"$label=$labelValue"))
-        .get(serviceName.toLowerCase, portName.toLowerCase)
-        .map(toNameTree(path, _))
+  private[this] val PrefixLen = 3
+  private[this] val variablePrefixLength = PrefixLen + labelName.size
 
-      Activity(nameTree.map(Activity.Ok(_)))
+  private[this] val servicesMemo =
+    Memoize[(String, String, Option[String]), Activity[Svc]] {
+      case (nsName, serviceName, labelSelector) =>
+        val eventLogger = EventLogger(nsName, serviceName)
+        mkApi(nsName)
+          .service(serviceName)
+          .activity(
+            Svc.fromResponse(_),
+            labelSelector = labelSelector
+          ) { case (svc, event) => svc.update(eventLogger, event) }
+    }
 
-    case _ =>
-      Activity.value(NameTree.Neg)
+  private[this] def service(
```

Comment:
Is there a purpose for this method?  Could callers just call `servicesMemo` instead?